File: packages/shared-types/wallet.ts
import type { IsoDatetimeString, JsonObject } from './json';

export enum WalletChain {
  ETHEREUM = 'ethereum',
  BSC = 'bsc',
  POLYGON = 'polygon',
  ARBITRUM = 'arbitrum',
  BASE = 'base',
  SOLANA = 'solana',
  TON = 'ton',
  TRON = 'tron',
}

export interface Wallet {
  id: string;
  userId: string;
  chain: WalletChain | string;
  address: string;
  isPrimary: boolean;
  verifiedAt?: Date;
  createdAt: Date;
}

export interface WalletWire {
  id: string;
  user_id: string;
  chain: WalletChain | string;
  address: string;
  is_primary: boolean;
  verified_at?: IsoDatetimeString;
  created_at: IsoDatetimeString;
}

export enum WchTransactionType {
  DEPOSIT = 'deposit',
  WITHDRAWAL = 'withdrawal',
  SUBSCRIPTION_PAYMENT = 'subscription_payment',
  CREDIT = 'credit',
  REWARD = 'reward',
  ADJUSTMENT = 'adjustment',
}

export enum WchTransactionStatus {
  PENDING = 'pending',
  CONFIRMED = 'confirmed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

export interface WchTransaction {
  id: string;
  userId: string;
  walletId?: string;
  txHash?: string;
  type: WchTransactionType | string;
  amount: number;
  status: WchTransactionStatus | string;
  metadata: JsonObject;
  createdAt: Date;
}

export interface WchTransactionWire {
  id: string;
  user_id: string;
  wallet_id?: string;
  tx_hash?: string;
  type: WchTransactionType | string;
  amount: number;
  status: WchTransactionStatus | string;
  metadata: JsonObject;
  created_at: IsoDatetimeString;
}

File: packages/go/domain/wallet.go
package domain

import "time"

type WalletChain string

const (
	WalletChainEthereum WalletChain = "ethereum"
	WalletChainBSC      WalletChain = "bsc"
	WalletChainPolygon  WalletChain = "polygon"
	WalletChainArbitrum WalletChain = "arbitrum"
	WalletChainBase     WalletChain = "base"
	WalletChainSolana   WalletChain = "solana"
	WalletChainTON      WalletChain = "ton"
	WalletChainTron     WalletChain = "tron"
)

type Wallet struct {
	ID         string      `json:"id"`
	UserID     string      `json:"user_id"`
	Chain      WalletChain `json:"chain"`
	Address    string      `json:"address"`
	IsPrimary  bool        `json:"is_primary"`
	VerifiedAt *time.Time  `json:"verified_at,omitempty"`
	CreatedAt  time.Time   `json:"created_at"`
}

type WchTransactionType string

const (
	WchTransactionTypeDeposit             WchTransactionType = "deposit"
	WchTransactionTypeWithdrawal          WchTransactionType = "withdrawal"
	WchTransactionTypeSubscriptionPayment WchTransactionType = "subscription_payment"
	WchTransactionTypeCredit              WchTransactionType = "credit"
	WchTransactionTypeReward              WchTransactionType = "reward"
	WchTransactionTypeAdjustment          WchTransactionType = "adjustment"
)

type WchTransactionStatus string

const (
	WchTransactionStatusPending   WchTransactionStatus = "pending"
	WchTransactionStatusConfirmed WchTransactionStatus = "confirmed"
	WchTransactionStatusFailed    WchTransactionStatus = "failed"
	WchTransactionStatusCancelled WchTransactionStatus = "cancelled"
)

type WchTransaction struct {
	ID        string                 `json:"id"`
	UserID    string                 `json:"user_id"`
	WalletID  *string                `json:"wallet_id,omitempty"`
	TxHash    *string                `json:"tx_hash,omitempty"`
	Type      WchTransactionType     `json:"type"`
	Amount    float64                `json:"amount"`
	Status    WchTransactionStatus   `json:"status"`
	Metadata  JSONObject             `json:"metadata"`
	CreatedAt time.Time              `json:"created_at"`
}
